[WIP] documentation: overflow rules in 'solidity in depth -> misc'

androlo opened the pull request:
A short section on overflow rules by the different operators, for signed and unsigned integers.

androlo committed:
overflow rules in 'solidity in depth -> misc'

axic commented:
I think there's another small section about overflow already (which explains what conditions can throw).

axic commented on docs/miscellaneous.rst:
```diff
@@ -303,6 +303,95 @@ The following is the order of precedence for operators, listed in order of evalu
++------------+----------+-----------+---------+
+
+1) Unary ``+`` has been deprecated.
+2) Due to the overflow protection for ``INT_MIN = 2**255``, which is built into the EVM, we have the following relation ``-INT_MIN == INT_MIN``. This affects a numer of arithmetic operations. See: `The Ethereum Yellow Paper <http://gavwood.com/paper>`_, ``SDIV``, Appendix H, Section 2 (Instruction set).
```
I think the YP has moved to https://ethereum.github.io/yellowpaper

axic commented on docs/miscellaneous.rst:
```diff
@@ -303,6 +303,95 @@ The following is the order of precedence for operators, listed in order of evalu
+2) Due to the overflow protection for ``INT_MIN = 2**255``, which is built into the EVM, we have the following relation ``-INT_MIN == INT_MIN``. This affects a numer of arithmetic operations. See: `The Ethereum Yellow Paper <http://gavwood.com/paper>`_, ``SDIV``, Appendix H, Section 2 (Instruction set).
+3) The exponent must be signed.
+4) Arithmetic left shift, implemented as truncated multiplication ``n << m == n * 2**m`` (overflowing bits are removed). The bitsize of the shifted value is that of ``n``.
+5) Arithmetic right shift, implemented as signed division ``n >> m := n / 2**m``. The bitsize of the shifted value is that of ``n``.
```
Is the `:=` supposed to be `==`?

androlo replied:
yes

androlo commented:
oh. yes.

On Tue, Oct 3, 2017 at 10:21 PM, Alex Beregszaszi <notifications@github.com>
wrote:

> *@axic* commented on this pull request.
> ------------------------------
>
> In docs/miscellaneous.rst
> <https://github.com/ethereum/solidity/pull/3017#discussion_r142511490>:
>
> > +| >>, >>=    |          |           | 5       |
> ++------------+----------+-----------+---------+
> +| &, &=      |          |           |         |
> ++------------+----------+-----------+---------+
> +| \|, \|=    |          |           |         |
> ++------------+----------+-----------+---------+
> +| ^, ^=      |          |           |         |
> ++------------+----------+-----------+---------+
> +| ~, ~=      |          |           |         |
> ++------------+----------+-----------+---------+
> +
> +1) Unary ``+`` has been deprecated.
> +2) Due to the overflow protection for ``INT_MIN = 2**255``, which is built into the EVM, we have the following relation ``-INT_MIN == INT_MIN``. This affects a numer of arithmetic operations. See: `The Ethereum Yellow Paper <http://gavwood.com/paper>`_, ``SDIV``, Appendix H, Section 2 (Instruction set).
> +3) The exponent must be signed.
> +4) Arithmetic left shift, implemented as truncated multiplication ``n << m == n * 2**m`` (overflowing bits are removed). The bitsize of the shifted value is that of ``n``.
> +5) Arithmetic right shift, implemented as signed division ``n >> m := n / 2**m``. The bitsize of the shifted value is that of ``n``.
>
> Is the := supposed to be ==?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/ethereum/solidity/pull/3017#pullrequestreview-66884805>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACremyFDNPVNrDV9m8uNpRV_8lwisXvzks5sopdIgaJpZM4Psm20>
> .
>



-- 
Regards // A

androlo commented:
seems there is also an error at `%` in signed numbers table.

edit: yeah, seems i forgot to move 'x' after i merged rows to put multiple related operations in the same row.

androlo commented:
hmm should probably add a note at signed `%` though, to let people know which implementation is used.

@axic I know there is a short section on truncation and some stuff, under 'solidity in depth -> types -> integers', but not much there about overflow. I still thought this would be more appropriate to put in misc tho, as an appendix.

androlo committed:
fixes to docs pr

androlo committed:
removed weird 'misc' file

androlo commented:
Latest with fixes in my repo. Diff: https://github.com/ethereum/solidity/commit/6b1fedafcadb7f4bc7044b80f69a11a027b0e3ce?short_path=55237a6#diff-55237a659404c779012062f2c78ea50d

androlo committed:
change headline

chriseth commented:
Could you add a short paragraph explaining the difference between (or perhaps even defining) overflow, underflow and truncation?

androlo committed:
added some info about over/underflow and truncation, and a note about exp 0**0 = 1

androlo committed:
Fixed typos

androlo commented:
Not very good at writing but I added a small section above the tables. Restructured it a little bit too, and changed some headlines.

redsquirrel commented on docs/miscellaneous.rst:
```diff
@@ -303,6 +303,115 @@ The following is the order of precedence for operators, listed in order of evalu
+Overflow and underflow rules for operators
+==========================================
+
+Integers in Solidity has a fixed range of possible values, and that range is defined by their bit-size. When an (arithmetic) operation attempts to create a number that is greater than the maximum value it results in an *overflow*. If it attempts to create a number that is less than the minimum value allowed then it is called an *underflow*.
```
Integers in Solidity _have_ a fixed range...

androlo replied:
i fixed

androlo commented:
@chriseth i will add definitions, and also distinctions between wrapping and over/underflow.

androlo committed:
A few changes to the terms used.

androlo commented:
I made a few more updates. 

1. Removed the `-INT_MIN = INT_MIN` note from the tables and gave it a special section in the introductory notes.
2. Removed the "truncation" term entirely, and instead use "clamped" when bits are being truncated. Don't want to mix it up with truncating decimal parts of floats.
3. Added a few lines about casting and demotion.
4. Added a warning that points out over/underflow is not detected. Maybe it should be more verbose, but that normally means that it is not checked for, or treated as an error. Also, the Z3 module seem to check for this at compile time; maybe that should be included as well.

androlo commented:
I could add that there are several related tests in the (extended) test suite for the libraries i'm working on - which also contains a simple Math library that has a few functions for doing exact arithmetic.

axic commented on docs/miscellaneous.rst:
```diff
@@ -303,6 +303,124 @@ The following is the order of precedence for operators, listed in order of evalu
+Overflow and underflow in signed and unsigned integers are generally handled by reducing the value ``modulo N``, where ``N`` is the bitsize of the integer in question, although there are some important exceptions (see the **special cases** section below).
+
+.. warning::
+Solidity does not automatically detect over/underflow.
```
I think this must be indented 4 spaces to be part of the warning block.

androlo replied:
ok i fixed.

androlo committed:
changed indentation of text at warning block

chriseth commented on docs/miscellaneous.rst:
```diff
@@ -303,6 +303,124 @@ The following is the order of precedence for operators, listed in order of evalu
+Overflow and underflow rules for operators
+==========================================
+
+Integers in Solidity have a fixed range of possible values, and that range is defined by their bit-size. When an (arithmetic) operation attempts to create a number that is greater than the maximum value it results in an *overflow*. If it attempts to create a number that is less than the minimum value allowed then it is called an *underflow*.
```
Can you wrap long lines, please? Changes to shorter lines are easier to view in diff tools.

androlo replied:
All lines will be using the same formatting as the other text in the document in the next push, so it should be fixed.

androlo replied:
pushed and checked.

chriseth commented on docs/miscellaneous.rst:
```diff
@@ -303,6 +303,124 @@ The following is the order of precedence for operators, listed in order of evalu
+
+Integers in Solidity have a fixed range of possible values, and that range is defined by their bit-size. When an (arithmetic) operation attempts to create a number that is greater than the maximum value it results in an *overflow*. If it attempts to create a number that is less than the minimum value allowed then it is called an *underflow*.
+
+Unsigned integers are restricted to the range ``[0, 2**N - 1]``, where ``N`` is the bitsize.
```
You write `bit-size` above and `bitsize` here.

androlo replied:
bitsize is used everywhere in next push.

androlo replied:
pushed and checked.

chriseth commented on docs/miscellaneous.rst:
```diff
@@ -303,6 +303,124 @@ The following is the order of precedence for operators, listed in order of evalu
+
+Unsigned integers are restricted to the range ``[0, 2**N - 1]``, where ``N`` is the bitsize.
+
+Signed integers are represented in two's complement, and are restricted to the range ``[-2**(N - 1), 2**(N - 1) - 1]``, , where ``N`` is the bitsize.
```
We could link `two's complement` to wikipedia.

androlo replied:
linked in next push.

androlo replied:
pushed and checked.

chriseth commented on docs/miscellaneous.rst:
```diff
@@ -303,6 +303,124 @@ The following is the order of precedence for operators, listed in order of evalu
+Overflow and underflow in signed and unsigned integers are generally handled by reducing the value ``modulo N``, where ``N`` is the bitsize of the integer in question, although there are some important exceptions (see the **special cases** section below).
+
+.. warning::
+    Solidity does not automatically detect over/underflow.
```
This should be part of the first paragraph.

androlo replied:
fixed in next push.

androlo replied:
pushed and checked.

chriseth commented on docs/miscellaneous.rst:
```diff
@@ -303,6 +303,124 @@ The following is the order of precedence for operators, listed in order of evalu
+.. warning::
+    Solidity does not automatically detect over/underflow.
+
+Additionally, the EVM has its own overflow semantics for results that exceed 256 bits. See: `The Ethereum Yellow Paper <https://ethereum.github.io/yellowpaper>`_, Appendix H, Section 2 (Instruction set).
```
Are they different from Solidity's or general two's complement?

androlo replied:
removed in next push. only confusing.

androlo replied:
pushed and checked.

chriseth commented on docs/miscellaneous.rst:
```diff
@@ -303,6 +303,124 @@ The following is the order of precedence for operators, listed in order of evalu
+
+Examples:
+
+``uint8(257) == 257 mod 2**8 == 1``
```
I would either use mathematical notation (`mod`, `=` and multi-ary comparisons) or Solidity (`%`, `==` and only binary comparisons) notation, but not mixed. Since mathematical notation is extremely cumbersome to read for novices especially in the area of modulos, I would prefer Solidity notation.

androlo replied:
changed to the arithmetic style you propose below. will be included in the next push. the modulo step is removed completely.

androlo replied:
just to clarify - by removing it there is no need for either math or a symbol which in this context would implicitly mean the exact thing as "math mod", but in reality does not (which is what the weird notation was trying to avoid in the first place). 

cleaning out a lot of this extra stuff everywhere. sorry, this is not a good pr.

chriseth commented on docs/miscellaneous.rst:
```diff
@@ -303,6 +303,124 @@ The following is the order of precedence for operators, listed in order of evalu
+
+**Casting and demotion**
+
+When casting a value into an integer type, the result is always reduced `modulo N`, where `N` is the bit-size of that integer type. This is important to remember when demoting an integer to a new type with a smaller bit-size.
```
double backticks

androlo replied:
fixed in next push

androlo replied:
pushed and checked.

chriseth commented on docs/miscellaneous.rst:
```diff
@@ -303,6 +303,124 @@ The following is the order of precedence for operators, listed in order of evalu
+
+When casting a value into an integer type, the result is always reduced `modulo N`, where `N` is the bit-size of that integer type. This is important to remember when demoting an integer to a new type with a smaller bit-size.
+
+Example: ``uint8(257) = 1``;
```
We had these examples already....

Perhaps better use arithmetics in the examples above? `uint8 x = 255; x + 2 == 1`

androlo replied:
fixed in next push.

androlo replied:
pushed and checked.

chriseth commented on docs/miscellaneous.rst:
```diff
@@ -303,6 +303,124 @@ The following is the order of precedence for operators, listed in order of evalu
+
+**Special cases**
+
+The range of signed integer types, i.e. ``[-2**(N - 1), 2**(N - 1) - 1]``, is asymmetric around 0. As a result of this, inverting the minimum value will result in a value that is ``1`` greater then the maximum value, making it wrap around back to itself:
```
Perhaps start with `In two's complement, the range...` - otherwise it sounds like a peculiarity of Solidity.

androlo replied:
fixed in next push.

androlo replied:
section removed since it's covered by wikipedia entry on two's complement. pushed and checked.

chriseth commented on docs/miscellaneous.rst:
```diff
@@ -303,6 +303,124 @@ The following is the order of precedence for operators, listed in order of evalu
+
+The range of signed integer types, i.e. ``[-2**(N - 1), 2**(N - 1) - 1]``, is asymmetric around 0. As a result of this, inverting the minimum value will result in a value that is ``1`` greater then the maximum value, making it wrap around back to itself:
+
+``-INT_MIN = -(-2**(N - 1)) = 2**(N - 1) = 2**(N - 1) - 1 + 1 = INT_MAX + 1 = INT_MIN``
```
Perhaps better to be concrete and not use non-introduced constants:
`int8 x = -127; -x == -127;`.

androlo replied:
removed completely. the wiki entry for two's complement (which is linked to now) explains this in the section about the most negative number.

androlo replied:
pushed and checked.

chriseth commented on docs/miscellaneous.rst:
```diff
@@ -303,6 +303,124 @@ The following is the order of precedence for operators, listed in order of evalu
++------------+----------+-----------+---------+
+|  Operator  | Overflow | Underflow | Notes   |
++============+==========+===========+=========+
+| +, ++, +=  | x        |           |         |
```
This can underflow: `int8 x = -120; x += - 100;`.

androlo replied:
my current table for signed integers says it can do both. maybe it is something wrong with the commits (i think something went wrong after copy pasting a few commits back).

chriseth commented on docs/miscellaneous.rst:
```diff
@@ -303,6 +303,124 @@ The following is the order of precedence for operators, listed in order of evalu
++------------+----------+-----------+---------+
+| \|, \|=    |          |           |         |
++------------+----------+-----------+---------+
+| ^, ^=      |          |           |         |
```
Strictly, this can also overflow or underflow. We just don't call it like that because it is a bit operator.

Perhaps it would be better to talk about the operators in paragraphs instead of a table?

androlo replied:
could use N/A for bit operators, to emphasize the difference. Btw I included arithmetic shift as an arithmetic operation since it is a form of hybrid.

I don't understand what paragraphs means. Does it mean removing the table and instead give every group of operators its own paragraph?

androlo replied:
this is another big issue, like introducing invariants. i will just leave this as-is for the next push, which i will do in a little while. I sense this will go down the arithmetic/bitwise rabbit hole 😧.

chriseth commented on docs/miscellaneous.rst:
```diff
@@ -303,6 +303,124 @@ The following is the order of precedence for operators, listed in order of evalu
++------------+----------+-----------+---------+
+
+1) Unary ``+`` has been deprecated.
+2) Unary ``-`` works on ``uints``, e.g. ``-uint(1) == ~uint(0)``.
```
_also_ works on uints.

chriseth replied:
Perhaps we should also tell users of invariants: `-x` is always identical to `0 - x`. `a += b` is always identical to `a = a + b` (with the difference that a is only evaluated once) and so on.

androlo replied:
first suggestion is fixed in the next push. 

i will push without any of those invariants, so that is the only remaining issue (since it seems like a pretty big one).

androlo committed:
First fixes for chriseth's comments

androlo committed:
fixes after chriseth's comments, additional cleanup

androlo committed:
fixes after chriseth's comments, additional cleanup

androlo committed:
fixes after chriseth's comments, additional cleanup

axic commented:
Can the commits be squashed a bit and rebased against`develop`?

androlo commented:
definitely. this is on hold tho because christian suggested tables are turned into paragraphs, and potentially there will also be a list of invariants included. also need to know if under/overflow in bit operations should be changed from "no" to "undefined" (all except shifts, since they are multiplication). 

Personally i am in favor of using undefined, which means there are three cases for each "flow direction":

1. yes: it is possible for the given operator with the given input.
2. no: it is not possible for the given operator with the given input.
3. undefined: it is not meaningful to talk about under/overflow with the given operator.

Undefined is only used for unary `+` now (since it's being removed).

axic commented:
Rebased this to avoid merge commits to ease the review process.

axic commented:
@androlo how about squashing some commits and merging this?

chriseth commented on docs/miscellaneous.rst:
```diff
@@ -303,6 +303,121 @@ The following is the order of precedence for operators, listed in order of evalu
++------------+----------+-----------+---------+
+| \*\*       | x        | x         | 2       |
++------------+----------+-----------+---------+
+| /, /=      |          |           |         |
```
This actually can overflow.

I'm not sure if a table is the best way to present this information. I would prefer text that also explains what overflow means in such a case, what actually happens and why some operators cannot overflow.

androlo replied:
When two signed 2's complement numbers are added, overflow is detected if:

- both operands are positive and the result is negative, or

- both operands are negative and the result is positive.

When two unsigned numbers are added, overflow occurs if

- there is a carry out of the leftmost bit.


Is that good? Just extend that to include multiplication etc.

androlo replied:
also, my table has the updated rules from having edited. this is an old pr i will remove it since it has been updated so much and is getting messy.

androlo commented:
this is old and has been edited a lot so i removed it.
